File: test/HaskellReactSpec.hs
{-# LANGUAGE PackageImports #-}

module HaskellReactSpec where

import FFI (Defined(Defined))
import HaskellReact
import "fay-base" Data.Text (Text, append, showInt, pack, unpack)
import Prelude hiding (span)
import Tag.Input (input, defaultInputAttributes, onChange)

data ReactState = ReactState {
  header1 :: Text
  , countClicks :: Int
}

data SimpleState = SimpleState { number :: Int }

render' :: ReactInstance ReactState -> Fay DOMElement
render' reactInstance = do
  data' <- state reactInstance
  let text = (header1 data') `append` (pack " ") `append` (showInt $ countClicks data')
      onClick = const $ setState reactInstance (data' { countClicks = countClicks data' + 1} )
  return $ constructDOMElement "a" (defaultAttributes { className = Defined "blue", onClick = Defined onClick }) text

singleElement :: DOMElement
singleElement = let
  innerData = ReactState (pack "The header") 0
  reactData = ReactData {
    render = render'
    , componentDidMount = return ()
    , displayName = "SpanClass"
    , getInitialState = innerData
  }
  in classInstance (declareReactClass reactData)

element :: DOMElement
element = let
  innerData = ReactState (pack "Element") 0
  reactData = (defaultReactData innerData) {
    render = \reactInstance -> do
      mounted <- isMounted reactInstance
      return $ constructDOMElement "h1" defaultAttributes (pack $ show mounted)
  }
  in classInstance (declareReactClass reactData)

aElement :: DOMElement
aElement = let
  innerData = ReactState (pack "AElement") 0
  reactData = (defaultReactData innerData) {
    render = \reactInstance -> return $ let
      aAttr = aAttributesDefaults {
        href = Defined $ pack $ "http://google.com"
      }
      in a defaultAttributes aAttr (pack "Google")
  }
  in classInstance (declareReactClass reactData)

relatedElements :: DOMElement
relatedElements = let
  reactData = (defaultReactData (SimpleState 0)) {
    render = \reactInstance -> do
      actualState <- state reactInstance
      let spanElement = span (pack "Num: " `append` (pack $ show $ number actualState))
          inputElement = input defaultAttributes (defaultInputAttributes {
            onChange = Defined $ \changeEvent -> do
              value <- eventValue changeEvent
              let ss = SimpleState $ length value
              (setState reactInstance ss)
          }) (pack "")
          divElement = constructDOMElementArray "div" defaultAttributes [spanElement, inputElement]
      return divElement
    }
  in declareAndRun reactData

File: test/haskellReactSpec.js
describe("Haskell React", function() {
  it("the click event works", function() {
    var singleElement = Fay$$_(HaskellReactSpec.singleElement, true);
    var ReactTestUtils = React.addons.TestUtils;
    var rendered = ReactTestUtils.renderIntoDocument(singleElement).getDOMNode();
    expect(rendered.innerText).toEqual("The header 0");
    ReactTestUtils.Simulate.click(rendered);
    ReactTestUtils.Simulate.click(rendered);
    expect(rendered.innerText).toEqual("The header 2");
  });
  it("component correctly displays isMounted", function () {
    var component = Fay$$_(HaskellReactSpec.element, true);
    var ReactTestUtils = React.addons.TestUtils;
    var rendered = ReactTestUtils.renderIntoDocument(component).getDOMNode();
    expect(rendered.innerText).toEqual("false");
  });
  it("dom node", function () {
    var component = Fay$$_(HaskellReactSpec.aElement, true);
    var ReactTestUtils = React.addons.TestUtils;
    var rendered = ReactTestUtils.renderIntoDocument(component).getDOMNode();
    expect(rendered.innerText).toEqual("Google");
    expect(rendered.getAttribute("href")).toEqual("http://google.com");
  });
  it("can link two elements", function () {
    var component = Fay$$_(HaskellReactSpec.relatedElements, true);
    var ReactTestUtils = React.addons.TestUtils;
    var rendered = ReactTestUtils.renderIntoDocument(component);
    var input = ReactTestUtils.findRenderedDOMComponentWithTag(rendered, "input");
    ReactTestUtils.Simulate.change(input.getDOMNode(), { target: { value: 'abcde' } });
    var span = ReactTestUtils.findRenderedDOMComponentWithTag(rendered, "span");
    expect(span.getDOMNode().innerText).toEqual("Num: 5");
  });
});
